devito/passes/clusters/cse.py

```diff
@@ -160,65 +231,181 @@ def _compact_temporaries(exprs, exclude):
     return processed
 
 
+def _toposort(exprs):
+    """
+    Ensure the expression list is topologically sorted.
+    """
+    dag = DAG(exprs)
+
+    for e0 in exprs:
+        if not isinstance(e0.lhs, CTemp):
+            continue
+
+        for e1 in exprs:
+            if e0.lhs in e1.rhs.free_symbols:
+                dag.add_edge(e0, e1, force_add=True)
+
+    def choose_element(queue, scheduled):
+        # Try to honor temporary names as much as possible
+        first = sorted(queue, key=lambda i: str(i.lhs)).pop(0)
+        queue.remove(first)
+        return first
+
+    processed = dag.topological_sort(choose_element)
+
+    return processed
+
+
+class Candidate(tuple):
+
+    def __new__(cls, expr, conditionals=None, sources=()):
+        conditionals = frozendict(conditionals or {})
+        sources = as_tuple(sources)
+        return tuple.__new__(cls, (expr, conditionals, sources))
+
+    @property
+    def expr(self):
+        return self[0]
+
+    @property
+    def conditionals(self):
+        return self[1]
+
+    @property
+    def sources(self):
+        return self[2]
+
+    @property
+    def key(self):
+        return (self.expr, self.conditionals)
+
+    @cached_property
+    def cost(self):
+        if len(self.sources) == 1:
+            return 0
+        else:
+            return estimate_cost(self.expr)
+
+    def as_subs(self, v):
+        subs = {self.expr: v}
+
+        # Also add in subs for compound-based replacement
+        # E.g., `a*b*c*d` -> `r0*c*d`
+        for i in self.sources:
+            if self.expr == i:
+                continue
+
+            args = [v]
+            queue = list(self.expr.args)
+            for a in i.args:
+                try:
+                    queue.remove(a)
+                except ValueError:
+                    args.append(a)
+            assert not queue
+            subs[i] = self.expr.func(*args)
+
+        return subs
+
+
+def catch(exprs, mode):
```

Comment:
I can't use `search` because it'd be an alias of `symbolics.search`, and `retrieve_cse` is misleading given what "cse" means and in particular given that we're just looking for _potential_ common subexpressions (that is, candidates)

I think `catch` (for Candidates) is fine